x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,507 @@
+	cfgwarn.Beta("s3 input type is used")
+	logger := logp.NewLogger(inputName)
+
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrap(err, "failed unpacking config")
+	}
+
+	out, err := connector.ConnectWith(cfg, beat.ClientConfig{
+		Processing: beat.ProcessingConfig{
+			DynamicFields: context.DynamicFields,
+		},
+		ACKEvents: func(privates []interface{}) {
+			for _, private := range privates {
+				if s3Context, ok := private.(*s3Context); ok {
+					s3Context.done()
+				}
+			}
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	awsConfig, err := awscommon.GetAWSCredentials(config.AwsConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "getAWSCredentials failed")
+	}
+
+	closeChannel := make(chan struct{})
+	p := &Input{
+		outlet:    out,
+		config:    config,
+		awsConfig: awsConfig,
+		logger:    logger,
+		close:     closeChannel,
+		context:   &channelContext{closeChannel},
+	}
+	return p, nil
+}
+
+// Run runs the input
+func (p *Input) Run() {
+	p.workerOnce.Do(func() {
+		p.workerWg.Add(1)
+		visibilityTimeout := int64(p.config.VisibilityTimeout.Seconds())
+		regionName, err := getRegionFromQueueURL(p.config.QueueURL)
+		if err != nil {
+			p.logger.Errorf("failed to get region name from queueURL: %v", p.config.QueueURL)
+		}
+
+		awsConfig := p.awsConfig.Copy()
+		awsConfig.Region = regionName
+		svcSQS := sqs.New(awsConfig)
+		svcS3 := s3.New(awsConfig)
+
+		go p.run(svcSQS, svcS3, visibilityTimeout)
+	})
+}
+
+func (p *Input) run(svcSQS *sqs.Client, svcS3 *s3.Client, visibilityTimeout int64) {
+	defer p.logger.Infof("s3 input worker for '%v' has stopped.", p.config.QueueURL)
+	defer p.workerWg.Done()
+	p.logger.Infof("s3 input worker has started. with queueURL: %v", p.config.QueueURL)
+	for p.context.Err() == nil {
+		// receive messages from sqs
+		req := svcSQS.ReceiveMessageRequest(
+			&sqs.ReceiveMessageInput{
+				QueueUrl:              &p.config.QueueURL,
+				MessageAttributeNames: []string{"All"},
+				MaxNumberOfMessages:   &maxNumberOfMessage,
+				VisibilityTimeout:     &visibilityTimeout,
+				WaitTimeSeconds:       &waitTimeSecond,
+			})
+
+		output, err := req.Send(p.context)
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == awssdk.ErrCodeRequestCanceled {
+				continue
+			}
+			p.logger.Error("failed to receive message from SQS:", err)
+			time.Sleep(time.Duration(waitTimeSecond) * time.Second)
+			continue
+		}
+
+		if output == nil || len(output.Messages) == 0 {
+			p.logger.Debug("no message received from SQS:", p.config.QueueURL)
+			continue
+		}
+
+		// process messages received from sqs, get logs from s3 and create events
+		p.processor(p.config.QueueURL, output.Messages, visibilityTimeout, svcS3, svcSQS)
+	}
+}
+
+// Stop stops the s3 input
+func (p *Input) Stop() {
+	defer p.outlet.Close()
+	close(p.close)
+	p.context.Done()
+	p.workerWg.Wait()
```

Comment:
I think you have to add `p.workerWg.Wait` to `Wait`.

Reply:
Done, also added `stopOnce` to make sure `Stop()` only get called once.